[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode

jon-chuang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114056
* __->__ #113374

---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each subgraph.

It may come in handy in the future.

jon-chuang committed:
[fx] fx.split_module inserts autocast & grad_mode

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx] fx.split_module inserts autocast & grad_mode"




[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each graph.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each graph.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each graph.

[ghstack-poisoned]

ezyang commented:
deferring to @wconstab

wconstab commented on torch/fx/passes/split_module.py:
```diff
@@ -194,6 +196,76 @@ def instantiate_node_partition_mapping(node):
+        if active_grad is not None:
+            grad_regions[active_grad].add(split_callback(node))
+
+        for a in active_autocasts:
```
would you add a test case that covers more than one / nested autocast regions?  (and possibly also grad region in the same graph as autocast region)

jon-chuang replied:
Yes, I thought of doing so and to properly test we need a custom model which returns intermediate outputs whose requires_grad / dtype should differ based on which layer it is in (corresponding to which ctx manager is active in a given layer). I'll add it.

jon-chuang replied:
Added a nested test with example output:

```python
---orig graph---
graph():
    %l_inputs_ : torch.Tensor [num_users=2] = placeholder[target=L_inputs_]
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net1_0 : [num_users=1] = call_module[target=L__self___net1_0](args = (%l_inputs_,), kwargs = {})
    %l__self___net1_1 : [num_users=1] = call_module[target=L__self___net1_1](args = (%l__self___net1_0,), kwargs = {})
    %l__self___net1_2 : [num_users=1] = call_module[target=L__self___net1_2](args = (%l__self___net1_1,), kwargs = {})
    %l__self___net1_3 : [num_users=1] = call_module[target=L__self___net1_3](args = (%l__self___net1_2,), kwargs = {})
    %l__self___net1_4 : [num_users=1] = call_module[target=L__self___net1_4](args = (%l__self___net1_3,), kwargs = {})
    %intermediates_1 : [num_users=2] = call_module[target=L__self___net1_5](args = (%l__self___net1_4,), kwargs = {})
    %_enter_autocast_1 : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net2_0 : [num_users=1] = call_module[target=L__self___net2_0](args = (%intermediates_1,), kwargs = {})
    %l__self___net2_1 : [num_users=1] = call_module[target=L__self___net2_1](args = (%l__self___net2_0,), kwargs = {})
    %l__self___net2_2 : [num_users=1] = call_module[target=L__self___net2_2](args = (%l__self___net2_1,), kwargs = {})
    %intermediates_2 : [num_users=1] = call_module[target=L__self___net2_3](args = (%l__self___net2_2,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast_1,), kwargs = {})
    %l__self___net3_0 : [num_users=1] = call_module[target=L__self___net3_0](args = (%l_inputs_,), kwargs = {})
    %l__self___net3_1 : [num_users=1] = call_module[target=L__self___net3_1](args = (%l__self___net3_0,), kwargs = {})
    %l__self___net3_2 : [num_users=1] = call_module[target=L__self___net3_2](args = (%l__self___net3_1,), kwargs = {})
    %l__self___net3_3 : [num_users=1] = call_module[target=L__self___net3_3](args = (%l__self___net3_2,), kwargs = {})
    %l__self___net3_4 : [num_users=1] = call_module[target=L__self___net3_4](args = (%l__self___net3_3,), kwargs = {})
    %outputs : [num_users=1] = call_module[target=L__self___net3_5](args = (%l__self___net3_4,), kwargs = {})
    %_exit_autocast_1 : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return (intermediates_1, intermediates_2, outputs)

---split graph---
graph():
    %l_inputs_ : torch.Tensor [num_users=2] = placeholder[target=L_inputs_]
    %submod_0 : [num_users=1] = call_module[target=submod_0](args = (%l_inputs_,), kwargs = {})
    %submod_1 : [num_users=2] = call_module[target=submod_1](args = (%submod_0,), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%submod_1, 0), kwargs = {})
    %getitem_1 : [num_users=1] = call_function[target=operator.getitem](args = (%submod_1, 1), kwargs = {})
    %submod_2 : [num_users=2] = call_module[target=submod_2](args = (%getitem, %l_inputs_), kwargs = {})
    %getitem_2 : [num_users=1] = call_function[target=operator.getitem](args = (%submod_2, 0), kwargs = {})
    %getitem_3 : [num_users=1] = call_function[target=operator.getitem](args = (%submod_2, 1), kwargs = {})
    %submod_3 : [num_users=1] = call_module[target=submod_3](args = (%getitem_2,), kwargs = {})
    return (getitem_1, getitem_3, submod_3)

---submod_0 graph---
graph():
    %l_inputs_ : torch.Tensor [num_users=1] = placeholder[target=l_inputs_]
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net1_0 : [num_users=1] = call_module[target=L__self___net1_0](args = (%l_inputs_,), kwargs = {})
    %l__self___net1_1 : [num_users=1] = call_module[target=L__self___net1_1](args = (%l__self___net1_0,), kwargs = {})
    %l__self___net1_2 : [num_users=1] = call_module[target=L__self___net1_2](args = (%l__self___net1_1,), kwargs = {})
    %l__self___net1_3 : [num_users=1] = call_module[target=L__self___net1_3](args = (%l__self___net1_2,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return l__self___net1_3

---submod_1 graph---
graph():
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net1_3 : [num_users=1] = placeholder[target=l__self___net1_3]
    %l__self___net1_4 : [num_users=1] = call_module[target=L__self___net1_4](args = (%l__self___net1_3,), kwargs = {})
    %l__self___net1_5 : [num_users=2] = call_module[target=L__self___net1_5](args = (%l__self___net1_4,), kwargs = {})
    %_enter_autocast_1 : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net2_0 : [num_users=1] = call_module[target=L__self___net2_0](args = (%l__self___net1_5,), kwargs = {})
    %l__self___net2_1 : [num_users=1] = call_module[target=L__self___net2_1](args = (%l__self___net2_0,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast_1,), kwargs = {})
    %_exit_autocast_1 : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return (l__self___net2_1, l__self___net1_5)

---submod_2 graph---
graph():
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %_enter_autocast_1 : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net2_1 : [num_users=1] = placeholder[target=l__self___net2_1]
    %l_inputs_ : torch.Tensor [num_users=1] = placeholder[target=l_inputs_]
    %l__self___net2_2 : [num_users=1] = call_module[target=L__self___net2_2](args = (%l__self___net2_1,), kwargs = {})
    %l__self___net2_3 : [num_users=1] = call_module[target=L__self___net2_3](args = (%l__self___net2_2,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast_1,), kwargs = {})
    %l__self___net3_0 : [num_users=1] = call_module[target=L__self___net3_0](args = (%l_inputs_,), kwargs = {})
    %l__self___net3_1 : [num_users=1] = call_module[target=L__self___net3_1](args = (%l__self___net3_0,), kwargs = {})
    %_exit_autocast_1 : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return (l__self___net3_1, l__self___net2_3)

---submod_3 graph---
graph():
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net3_1 : [num_users=1] = placeholder[target=l__self___net3_1]
    %l__self___net3_2 : [num_users=1] = call_module[target=L__self___net3_2](args = (%l__self___net3_1,), kwargs = {})
    %l__self___net3_3 : [num_users=1] = call_module[target=L__self___net3_3](args = (%l__self___net3_2,), kwargs = {})
    %l__self___net3_4 : [num_users=1] = call_module[target=L__self___net3_4](args = (%l__self___net3_3,), kwargs = {})
    %l__self___net3_5 : [num_users=1] = call_module[target=L__self___net3_5](args = (%l__self___net3_4,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return l__self___net3_5

```

wconstab replied:
>                 with self.ctx_manager_1():
                    intermediates_1 = self.net1(inputs)
                    with self.ctx_manager_2():
                        intermediates_2 = self.net2(intermediates_1)
                    outputs = self.net3(inputs)


~~Assuming the above graph is from this model, isn't it wrong? I would expect the exits to be in reverse order to the enters~~

ok nvm they are correct, but, confusingly the `_exit_autocast` node exits the `_enter_autocast_1`, while exit1 exits enter0...  Not sure if that's worth fixing (would it be hard?)

`    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast_1,), kwargs = {})
`

`    %_exit_autocast_1 : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
`

jon-chuang replied:
Yes, this is just how the fx graph names things when there are name collisions. I don't think it's worth fixing though it is confusing.

jon-chuang replied:
I'm going to resolve this - it's just an artifact of fx name collisions and is IMO not a useability concern as long as one scruitinizes the graph properly.

wconstab commented on torch/fx/passes/split_module.py:
```diff
@@ -207,14 +279,22 @@ def instantiate_node_partition_mapping(node):
             continue
 
-        instantiate_node_partition_mapping(node)
+        if assert_monotonically_increasing:
```
this is to enforce the graph is split into sequential chunks of nodes, rather than mapping some 'earlier' nodes to 'later' partitions?

jon-chuang replied:
Yes. Without this property we cannot be assured about which partitions will covered by a given enter and exit.

wconstab commented on torch/fx/passes/split_module.py:
```diff
@@ -207,14 +279,22 @@ def instantiate_node_partition_mapping(node):
-        )  # noqa: B950
+        # do not capture cross-partition dependencies for global state nodes as they will be
+        # self-contained - their setup and unwind will be isolated to each partition submodule.
+        if node.target not in GLOBAL_STATE_NODES:
```
i'm not clear on the reasons for needing this.

1- is there a 'use' relationship between set_grad_enabled node and some other fx nodes that come after it in the graph?
2- when we duplicate `set_grad_enabled` node at the beginning of other subgraphs, wouldn't we update the graph so that nodes in partition B only depend on partition B's `set_grad_enabled`?

given that supposition, i might claim we can safely record_cross_partition_use for all nodes.  (perhaps, we need to move that 'prelude' section up above record_cross_partition_use first though?

jon-chuang replied:
The particular thing this handles is that the way that the autocast._exit is implemented as passing the result of autocast._enter to it. This introduces a cross-partition dependency. However, downstream compilers neither know how to deal with passing the autocast object as an argument, nor handling the case where autocast state is mutated by the graph. 

That being said, I think we don't actually need to forbid cross-partition the arguments to autocast._enter and set_grad_enabled - as much as they are likely to be constants anyway. 

There could be a subtle correctness issue if we don't handle the case where they are not constants I.e. data dependent - and the value originates in some other subgraph. I'm not sure if this is possible to create a test case for as dynamo might enforce that these arguments are coerced into constants by guarding, but I could try to force the value to be a symint.

jon-chuang replied:
One such way would be to create a boolean tensor with 1 elem and call .item on it

jon-chuang replied:
Actually, let's not add a test. Looking a the dynamo code, there is no possibility of the args being non-constant.

jon-chuang replied:
I've updated to only exclude autocast._exit

wconstab replied:
i think i still dont get why exit depends on enter from another graph.  if you are in case 2 (need to add both enter and exit to a middle graph) then your new exit would depend only on your new enter, right?

and in case 3, the tail graph, where you add a new enter and there was an existing exit - wouldn't you update the old exit to reference the new enter, or, create a new exit too?

my point was that there shouldn't be any cross-graph deps related to autocast/amp after your work is done, so, the cross-graph validation should just be allowed to run normally.

jon-chuang replied:
Right, the problem is that the per-subgraph setup/unwind is not instantiated until later. We are still at the preprocessing stage. If we did include autocast._exit in the dependency recording, then we wrongly identify a dependency between the partition where it is setup and the partition where it is unwound.

wconstab replied:
yea thats what i meant by this comment

> (perhaps, we need to move that 'prelude' section up above record_cross_partition_use first though?

e.g. maybe we should first install the new enters/exits, then do teh dependency analysis.

However i'm not sure it matters.  Maybe its just fine either way.

jon-chuang replied:
> e.g. maybe we should first install the new enters/exits, then do teh dependency analysis.

Yeah, I don't think we want that, as the record cross-partition use is a preprocessing stage, similar to how the global state nodes paritition dependency info gathering is a preprocessing stage, where we collect information about how we want to transform the graph in the graph transformation stage.

jon-chuang replied:
I'm going to call this resolved, basically, we do this as we take a different path for preprocessing the dependencies for autocast._exit

wconstab commented on torch/fx/passes/split_module.py:
```diff
@@ -235,6 +315,23 @@ def instantiate_node_partition_mapping(node):
+                    kwargs={},
+                    type_expr=node.type,
+                )
+                new_node.meta = node.meta.copy()  # is it really a good idea to copy this?
```
what does this typically contain?

what are the alternatives to copying it?
 - leaving it empty
 - adding a tainted copy, including some metadata that effectively says "i was created by split_module"
 - ...?

jon-chuang replied:
Well meta is just a dict that can be used for anything. When the result is a Tensor, it often can store an example_value for instance.

Since we are dealing only with autocast/grad_enabled, I don't think we ought to be too worried, but let me inspect the metadata and see if there's anything that needs to be updated.

jon-chuang replied:
There is prior art for copying the metadata when creating the new graph splits I.e. line 386, but my worry is more about the fact that we are here not copying 1-1 as per prior art but duplicating existing nodes multiple times across the subgraphs.

jon-chuang replied:
Hmm, currently, the `meta` have the following values:

```
META PRELUDE {'creation_timestamp': 0}
META PRELUDE {'creation_timestamp': 0}
META PRELUDE {'creation_timestamp': 0}
META PRELUDE {'creation_timestamp': 0}
META EPILOGUE {'creation_timestamp': 0}
META EPILOGUE {'creation_timestamp': 0}
META EPILOGUE {'creation_timestamp': 0}
META EPILOGUE {'creation_timestamp': 0}
```

We may want to just update this for correctness sake - we can set it to be the creation timestamp of the first node in the partition.

jon-chuang replied:
Actually creation_timestamp is only consumed within dynamo, so I think we don't need to care. (it's used for tracing checkpoint restoration)

jon-chuang replied:
Created a PR to ensure this metadata lifetime is managed properly: https://github.com/pytorch/pytorch/pull/113517

jon-chuang commented on torch/fx/passes/split_module.py:
```diff
@@ -289,6 +386,24 @@ def instantiate_node_partition_mapping(node):
 
+    # Exit epilogue
+    for regions_mapping in [autocast_regions]:
+        for node in reversed(regions_mapping):
```
Add a comment that we need to unwind in reverse order to setting up.

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each graph.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each graph.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each graph.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each graph.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each graph.

[ghstack-poisoned]

jon-chuang commented on test/distributed/test_dynamo_distributed.py:
```diff
@@ -565,6 +570,163 @@ def opt_fn(inputs):
+
+        ctx_managers = [
+            (
+                lambda: torch.autocast(torch.device(self.device).type, torch.float16),
```
Hmm, could you change one of the ctx managers to be bfloat16?

jon-chuang replied:
Added  in follow up PR

jon-chuang commented on torch/fx/passes/split_module.py:
```diff
@@ -289,6 +386,25 @@ def instantiate_node_partition_mapping(node):
+                partition = partitions[str(r)]
+                exit_node = autocast_exits[node]
+                assert exit_node is not None, "Missing exit node"
+                new_node = partition.graph.create_node(
```
In particular @wconstab if we try to record cross partition dependencies on the old graph rather than new graph, this replacement of the exit node's argument to the subgraph's copy of the enter is not performed.

yf225 commented:
@jon-chuang Will this approach still work, if we need to delay the graph-splitting and the compilation to the first actual run of the model? (Please see https://github.com/pytorch/pytorch/issues/113812 for details)

jon-chuang commented:
@yf225 as discussed in https://github.com/pytorch/pytorch/pull/112231#issuecomment-1782940758, if graph splitting is delayed to post-autograd (i.e. inductor), it should be fine.

jon-chuang commented:
>  DDPOptimizer will still split the Dynamo graph, just that we do it at runtime (in the 1st iteration), instead of doing it at compile time.

I find this very hard to understand. Dynamo trace will also occur on first iteration of running the model. So how is this different from the status quo?

yf225 commented:
> I find this very hard to understand. Dynamo trace will also occur on first iteration of running the model. So how is this different from the status quo?

For fixing the Inductor output strides bug: on first iteration of running the model, we will have the real tensors with the correct strides, so we will be able to propogate the correct strides to all subgraphs created by DDPOptimizer.

For fixing autocast & grad_mode (this PR): I personally don't know if there is anything different on the metadata side, if we do graph-splitting at runtime instead of compile time. Still need to investigate.

Maybe a good action item is to merge this PR first, and then I will look at delaying DDPOptimizer splitting to first iteration.

jon-chuang commented:
> Maybe a good action item is to merge this PR first, and then I will look at delaying DDPOptimizer splitting to first iteration.

Yep, let's merge this first. I don't completely understand the plan to delay the graph splitting, so let's keep the discussion to that issue.

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each subgraph.

It may come in handy in the future.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each subgraph.

It may come in handy in the future.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each subgraph.

It may come in handy in the future.

[ghstack-poisoned]

jon-chuang commented:
@wconstab I think this and stacked PR should be in mergeable condition now

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each subgraph.

It may come in handy in the future.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each subgraph.

It may come in handy in the future.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each subgraph.

It may come in handy in the future.

[ghstack-poisoned]

jon-chuang committed:
Update on "[fx, DDP] fx.split_module will setup/unwind autocast & grad_mode"


---


Replaces: https://github.com/pytorch/pytorch/pull/112231
Fixes: https://github.com/pytorch/pytorch/issues/111794

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

---

See before and after this PR for the split fx modules here (for autocast mode): https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804274605


---

### Discussion 
We don't actually have to do this for grad mode: https://github.com/pytorch/pytorch/pull/112231#issuecomment-1804280031. It's not wrong to do it anyway, but maybe unnecessary? But may still be better to keep this PR's changes so we're sure what the grad mode state ought to be for each subgraph.

It may come in handy in the future.

[ghstack-poisoned]
